drivers/i3c/i3c_dw.c

```diff
@@ -0,0 +1,2349 @@
+
+	/* lookup known pids */
+	const struct i3c_device_id i3c_id = I3C_DEVICE_ID(pid);
+	struct i3c_device_desc *target = i3c_device_find(dev, &i3c_id);
+
+	if (target == NULL) {
+		LOG_INF("%s: PID 0x%012llx is not in registered device "
+			"list, given DA 0x%02x",
+			dev->name, pid, dyn_addr);
+	} else {
+		target->dynamic_addr = dyn_addr;
+		tmp = sys_read32(config->regs + DEV_CHAR_TABLE_LOC3(data->dctstartaddr, pos));
+		target->bcr = DEV_CHAR_TABLE_BCR(tmp);
+		target->dcr = DEV_CHAR_TABLE_DCR(tmp);
+
+		LOG_DBG("%s: PID 0x%012llx assigned dynamic address 0x%02x", dev->name, pid,
+			dyn_addr);
+	}
+	i3c_addr_slots_mark_i3c(&data->common.attached_dev.addr_slots, dyn_addr);
+
+	return 0;
+}
+
+/**
+ * @brief Perform Dynamic Address Assignment.
+ *
+ * @see i3c_do_daa
+ *
+ * @param dev Pointer to controller device driver instance.
+ *
+ * @return @see i3c_do_daa
+ */
+static int dw_i3c_do_daa(const struct device *dev)
+{
+	const struct dw_i3c_config *config = dev->config;
+	struct dw_i3c_data *data = dev->data;
+	struct dw_i3c_xfer *xfer = &data->xfer;
+	struct dw_i3c_cmd *cmd;
+	uint32_t olddevs, newdevs;
+	uint8_t p, last_addr = 0;
+	int32_t pos, addr, ret;
+	uint32_t present_state;
+
+	present_state = sys_read32(config->regs + PRESENT_STATE);
+	if (!(present_state & PRESENT_STATE_CURRENT_MASTER)) {
+		return -EACCES;
+	}
+
+	olddevs = ~(data->free_pos);
+
+	/* Prepare DAT before launching DAA. */
+	for (pos = 0; pos < data->maxdevs; pos++) {
+		if (olddevs & BIT(pos)) {
+			continue;
+		}
+
+		addr = i3c_addr_slots_next_free_find(&data->common.attached_dev.addr_slots,
+						     last_addr + 1);
+		if (addr == 0) {
+			return -ENOSPC;
+		}
+
+		p = odd_parity(addr);
+		last_addr = addr;
+		addr |= (p << 7);
+		sys_write32(DEV_ADDR_TABLE_DYNAMIC_ADDR(addr),
+			    config->regs + DEV_ADDR_TABLE_LOC(data->datstartaddr, pos));
+	}
+
+	pos = get_free_pos(data->free_pos);
+	if (pos < 0) {
+		LOG_ERR("%s: find free pos failed", dev->name);
+		return -ENOSPC;
+	}
+
+	ret = k_mutex_lock(&data->mt, K_MSEC(1000));
+	if (ret) {
+		LOG_ERR("%s: Mutex err (%d)", dev->name, ret);
+		return ret;
+	}
+	memset(xfer, 0, sizeof(struct dw_i3c_xfer));
+
+	xfer->ncmds = 1;
+	xfer->ret = -1;
+
+	cmd = &xfer->cmds[0];
+	cmd->cmd_hi = COMMAND_PORT_TRANSFER_ARG;
+	cmd->cmd_lo = COMMAND_PORT_TOC | COMMAND_PORT_ROC |
+		      COMMAND_PORT_DEV_COUNT(data->maxdevs - pos) | COMMAND_PORT_DEV_INDEX(pos) |
+		      COMMAND_PORT_CMD(I3C_CCC_ENTDAA) | COMMAND_PORT_ADDR_ASSGN_CMD;
+
+	start_xfer(dev);
+	ret = k_sem_take(&data->sem_xfer, K_MSEC(1000));
+	if (ret) {
+		LOG_ERR("%s: Semaphore err (%d)", dev->name, ret);
+		k_mutex_unlock(&data->mt);
+		return ret;
+	}
+
+	k_mutex_unlock(&data->mt);
+
+	if (data->maxdevs == cmd->rx_len) {
+		newdevs = 0;
+	} else {
+		newdevs = GENMASK(data->maxdevs - cmd->rx_len - 1, 0);
+	}
+	newdevs &= ~olddevs;
+
+	for (pos = find_lsb_set(newdevs); pos <= find_msb_set(newdevs); pos++) {
+		if (newdevs & BIT(pos)) {
```

Comment:
No? It's been so long since I wrote this.... the PR is over a year old now, I did a quick skim through the databook again.

What makes you think it should be `pos-1`?

Reply:
Please see the explanation of "find_lsb_set" function:
https://elixir.bootlin.com/zephyr/v4.0.0/source/include/zephyr/arch/common/ffs.h#L41

It says, "This routine finds the first bit set starting from the least significant bit
 * in the argument passed in and returns the index of that bit.  **Bits are
 * numbered starting at 1 from the least significant bit.  A return value of
 * zero indicates that the value passed is zero."**

Reply:
ah, i see. done